File: leetcode-codes/1121-partition-array-for-maximum-sum/solution.py
class Solution(object):
    def maxSumAfterPartitioning(self, arr, k):
        self.dp=[-1]*(len(arr)+1)
        return self.dfs(0,arr,k)
    
    def dfs(self,i,arr,k):
        if i==len(arr):
            return 0
        if self.dp[i]!=-1:
            return self.dp[i]
        
        _max ,maxsum = 0,0

        for j in range(1,k+1):
            if i+j<=len(arr):
                _max = max(_max,arr[i+j-1])
                maxsum = max(maxsum,_max*j+self.dfs(i+j,arr,k))

        self.dp[i]=maxsum
        return self.dp[i]
        

# Example usage:
arr = [1, 15, 7, 9, 2, 5, 10]
k = 3
s= Solution()
print(s.maxSumAfterPartitioning(arr, k))  # Output: 84

File: leetcode-codes/1121-partition-array-for-maximum-sum/solution.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long
class Solution {
public:
    int n,K;
    vector<ll>dp;
    int maxSumAfterPartitioning(vector<int>& arr, int k) {
        n = arr.size();
        K= k;
        dp = vector<ll>(n,-1);
        return dfs(0,arr);
    }
    ll dfs(int cur,vector<int>& arr){
        if(cur>=n)return 0;
        ll &cache = dp[cur];
        if(cache!=-1)
            return cache;
        ll _max = 0;
        for(int i =1;i<=K;i++){
            if(cur+i<=n){
                _max = max((ll)arr[cur+i-1],_max);
                cache = max(cache,_max*i+dfs(cur+i,arr));
            }
        }
        return cache;
    }
};
